### PR TITLE
[log_processor] Refactor `LogParser` to return errors

### PR DESCRIPTION
**Pull request by alxarch**
## Background

Returning errors from `LogParser.Parse()` will allow easier testing for parsers (on non valid log input) and centralizes parse error handling in the classifier, decoupling parsers from zap logger 

## Changes

- Change `LogParser#Parse` interface method signature from `Parse(log string) []*PantherLog` to `Parse(log string) ([]*PantherLog, error)`
- Change all implemented parsers to the new signature, returning the errors instead of logging
- Log the error in `classification.Classifier.safeLogParse`
- Modify `parsers/testutil/EqualPantherLog` to check for errors
- Update all tests to the new signature

## Testing

- Tested with `mage test:ci`

### CODE REVIEW

File: internal/log_analysis/log_processor/classification/classifier_test.go [hunk 1]
```diff
@@ -38,13 +38,13 @@ func (m *mockParser) New() parsers.LogParser {
 	return m // pass through (not stateful)
 }
 
-func (m *mockParser) Parse(log string) []*parsers.PantherLog {
+func (m *mockParser) Parse(log string) ([]*parsers.PantherLog, error) {
 	args := m.Called(log)
 	result := args.Get(0)
 	if result == nil {
-		return nil
+		return nil, nil
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/classification/classifier_test.go:45]: Method should be returning the "real" error back. Something like; 

```
func (m *mockParser) Parse(log string) ([]*parsers.PantherLog, error) {
 args := m.Called(log)
 result := args.Get(0)
 err := args.Error(1)
 if result == nil {
  return nil, err
 }
 return result.([]*parsers.PantherLog), err

```

File: internal/log_analysis/log_processor/parsers/awslogs/alb.go [hunk 2]
```diff
@@ -87,28 +88,24 @@ func (p *ALBParser) New() parsers.LogParser {
...
 	if err != nil {
-		zap.L().Debug("failed to parse the log as csv")
-		return nil
+		return nil, err
 	}
 
 	if len(record) < albMinNumberOfColumns {
-		zap.L().Debug("failed to parse the log as csv (wrong number of columns)")
-		return nil
+		return nil, err
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/parsers/awslogs/alb.go:98]: `err` will be nil. You need to return an error with custom message.

File: internal/log_analysis/log_processor/parsers/awslogs/aurora_mysql_audit.go [hunk 3]
```diff
@@ -59,28 +57,28 @@ type AuroraMySQLAuditParser struct {
...
 	if err != nil {
-		zap.L().Debug("failed to parse the log as csv")
-		return nil
+		return nil, err
 	}
 
 	if len(record) < auroraMySQLAuditMinNumberOfColumns {
-		zap.L().Debug("failed to parse the log as csv (wrong number of columns)")
-		return nil
+		return nil, err
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/parsers/awslogs/aurora_mysql_audit.go:76]: `err` will be nil, we need to populate it.

File: internal/log_analysis/log_processor/parsers/awslogs/vpc_flow_test.go [hunk 4]
```diff
@@ -143,6 +144,7 @@ func TestVpcFlowLogType(t *testing.T) {
 func checkVPCFlowLog(t *testing.T, header, log string, expectedEvent *VPCFlow) {
 	expectedEvent.SetEvent(expectedEvent)
 	parser := (&VPCFlowParser{}).New() // important to call New() to initialize reader
-	parser.Parse(header)
-	testutil.EqualPantherLog(t, expectedEvent.Log(), parser.Parse(log))
+	_, _ = parser.Parse(header)
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/parsers/awslogs/vpc_flow_test.go:147]: Probably better to check the result here (even though the code previously was not doing it). 
We should verify that the parser returns an empty array and no error.

File: internal/log_analysis/log_processor/parsers/nginxlogs/access.go [hunk 5]
```diff
@@ -56,46 +55,46 @@ type Access struct {
...
+var _ parsers.LogParser = (*AccessParser)(nil)
+
 func (p *AccessParser) New() parsers.LogParser {
 	return &AccessParser{}
 }
 
 // Parse returns the parsed events or nil if parsing failed
-func (p *AccessParser) Parse(log string) []*parsers.PantherLog {
+func (p *AccessParser) Parse(log string) ([]*parsers.PantherLog, error) {
 	reader := csv.NewReader(strings.NewReader(log))
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/parsers/nginxlogs/access.go:66]: Huh, this seems like a leftover. We should be using the streaming CSV parser (like we do in AWS ALB logs). If it's not too much of a hustle maybe you can do in this PR or we can do in a subsequent PR.

File: web/src/constants.ts [hunk 6]
```diff
@@ -103,6 +103,12 @@ export const LOG_TYPES = [
   'Suricata.DNS',
   'Syslog.RFC3164',
   'Syslog.RFC5424',
+  'GitLab.Rails',
+  'GitLab.API',
+  'GitLab.Audit',
+  'GitLab.Integrations',
+  'GitLab.Exceptions',
+  'GitLab.Git',
```
Comment (kostaspap) [inline web/src/constants.ts:111]: This PR doesn't add new parsers. Merge conflict?

File: internal/log_analysis/log_processor/destinations/s3_test.go [hunk 7]
```diff
@@ -67,13 +67,15 @@ type mockParser struct {
...
 
-func (m *mockParser) Parse(log string) []*parsers.PantherLog {
+var _ parsers.LogParser = (*mockParser)(nil)
+
+func (m *mockParser) Parse(log string) ([]*parsers.PantherLog, error) {
 	args := m.Called(log)
 	result := args.Get(0)
 	if result == nil {
-		return nil
+		return nil, nil
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/destinations/s3_test.go:76]: We should return the error returned by the parser. 
Should be changed to: 
```
args := m.Called(log)
result := args.Get(0)
err := args.Error(1)
if result == nil {
  return nil, err
}
return result.([]*parsers.PantherLog),err

```

File: web/src/constants.ts (hunk 6 above)
Comment (alxarch) [reply web/src/constants.ts]: Yep, I have a conflict in my master branch. Fixed

File: internal/log_analysis/log_processor/destinations/s3_test.go (hunk 7 above)
Comment (alxarch) [reply internal/log_analysis/log_processor/destinations/s3_test.go:76]: Fixed

File: internal/log_analysis/log_processor/parsers/nginxlogs/access.go (hunk 5 above)
Comment (alxarch) [reply internal/log_analysis/log_processor/parsers/nginxlogs/access.go:66]: I'd prefer a separate PR unless it's already checnged in upstream and the merge reverted it

Comment (kostaspap) [reply internal/log_analysis/log_processor/parsers/nginxlogs/access.go:66]: Fine by me. No, this looks like a miss upstream.

**Pull request merged**